File: apps/ai-service/src/api/schemas/portfolio.py
"""Portfolio API Pydantic schemas."""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


# ==========================================
# Portfolio Settings Schemas (Singleton data)
# ==========================================


class HeroData(BaseModel):
    """Hero section data."""

    name: str
    title: str
    tagline: str
    avatar: Optional[str] = None
    location: Optional[str] = None


class AboutData(BaseModel):
    """About section data."""

    summary: str
    highlights: list[str] = []


class EducationData(BaseModel):
    """Education section data."""

    school: str
    degree: str
    period: str
    gpa: Optional[str] = None
    rank: Optional[str] = None
    coursework: list[str] = []


class LifestyleData(BaseModel):
    """Lifestyle section data."""

    music: dict[str, Any] = {}
    routines: list[str] = []


class SocialData(BaseModel):
    """Social links data."""

    github: Optional[str] = None
    linkedin: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    twitter: Optional[str] = None
    website: Optional[str] = None


class PortfolioSettingResponse(BaseModel):
    """Portfolio setting response."""

    id: int
    key: str
    value: dict[str, Any]
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==========================================
# Experience Schemas
# ==========================================


class ExperienceCreate(BaseModel):
    """Create experience entry."""

    company: str = Field(..., min_length=1, max_length=100)
    role: str = Field(..., min_length=1, max_length=100)
    period: str = Field(..., min_length=1, max_length=50)
    highlights: list[str] = []
    display_order: int = 0
    is_active: bool = True


class ExperienceUpdate(BaseModel):
    """Update experience entry."""

    company: Optional[str] = None
    role: Optional[str] = None
    period: Optional[str] = None
    highlights: Optional[list[str]] = None
    display_order: Optional[int] = None
    is_active: Optional[bool] = None


class ExperienceResponse(BaseModel):
    """Experience response."""

    id: int
    company: str
    role: str
    period: str
    highlights: list[str]
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==========================================
# Tech Stack Schemas
# ==========================================


class TechStackCreate(BaseModel):
    """Create tech stack item."""

    name: str = Field(..., min_length=1, max_length=50)
    icon: str = Field(..., min_length=1, max_length=50)
    category: str = Field(..., min_length=1, max_length=50)  # language, ai, backend, frontend, database, devops
    display_order: int = 0
    is_active: bool = True


class TechStackUpdate(BaseModel):
    """Update tech stack item."""

    name: Optional[str] = None
    icon: Optional[str] = None
    category: Optional[str] = None
    display_order: Optional[int] = None
    is_active: Optional[bool] = None


class TechStackResponse(BaseModel):
    """Tech stack response."""

    id: int
    name: str
    icon: str
    category: str
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==========================================
# Project Schemas
# ==========================================


class ProjectCreate(BaseModel):
    """Create project entry."""

    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1)
    tags: list[str] = []
    image: Optional[str] = None
    link: Optional[str] = None
    github: Optional[str] = None
    is_featured: bool = False
    display_order: int = 0
    is_active: bool = True


class ProjectUpdate(BaseModel):
    """Update project entry."""

    title: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[list[str]] = None
    image: Optional[str] = None
    link: Optional[str] = None
    github: Optional[str] = None
    is_featured: Optional[bool] = None
    display_order: Optional[int] = None
    is_active: Optional[bool] = None


class ProjectResponse(BaseModel):
    """Project response."""

    id: int
    title: str
    description: str
    tags: list[str]
    image: Optional[str]
    link: Optional[str]
    github: Optional[str]
    is_featured: bool
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==========================================
# Publication Schemas
# ==========================================


class PublicationCreate(BaseModel):
    """Create publication entry."""

    title: str = Field(..., min_length=1, max_length=500)
    venue: str = Field(..., min_length=1, max_length=200)
    doi: Optional[str] = None
    year: int = Field(..., ge=1900, le=2100)
    display_order: int = 0
    is_active: bool = True


class PublicationUpdate(BaseModel):
    """Update publication entry."""

    title: Optional[str] = None
    venue: Optional[str] = None
    doi: Optional[str] = None
    year: Optional[int] = Field(default=None, ge=1900, le=2100)
    display_order: Optional[int] = None
    is_active: Optional[bool] = None


class PublicationResponse(BaseModel):
    """Publication response."""

    id: int
    title: str
    venue: str
    doi: Optional[str]
    year: int
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==========================================
# Achievement Schemas
# ==========================================


class AchievementCreate(BaseModel):
    """Create achievement entry."""

    title: str = Field(..., min_length=1, max_length=200)
    event: str = Field(..., min_length=1, max_length=200)
    organization: str = Field(..., min_length=1, max_length=200)
    year: int = Field(..., ge=1900, le=2100)
    display_order: int = 0
    is_active: bool = True


class AchievementUpdate(BaseModel):
    """Update achievement entry."""

    title: Optional[str] = None
    event: Optional[str] = None
    organization: Optional[str] = None
    year: Optional[int] = Field(default=None, ge=1900, le=2100)
    display_order: Optional[int] = None
    is_active: Optional[bool] = None


class AchievementResponse(BaseModel):
    """Achievement response."""

    id: int
    title: str
    event: str
    organization: str
    year: int
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==========================================
# Course Schemas
# ==========================================


class CourseCreate(BaseModel):
    """Create course entry."""

    title: str = Field(..., min_length=1, max_length=200)
    year: int = Field(..., ge=1900, le=2100)
    focus: list[str] = []
    display_order: int = 0
    is_active: bool = True


class CourseUpdate(BaseModel):
    """Update course entry."""

    title: Optional[str] = None
    year: Optional[int] = Field(default=None, ge=1900, le=2100)
    focus: Optional[list[str]] = None
    display_order: Optional[int] = None
    is_active: Optional[bool] = None


class CourseResponse(BaseModel):
    """Course response."""

    id: int
    title: str
    year: int
    focus: list[str]
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==========================================
# Reorder Schema
# ==========================================


class ReorderRequest(BaseModel):
    """Reorder items request."""

    ids: list[int]  # List of IDs in the desired order


# ==========================================
# Full Portfolio Response (Public API)
# ==========================================


class PortfolioFullResponse(BaseModel):
    """Complete portfolio data for public API."""

    hero: dict[str, Any]
    about: dict[str, Any]
    education: dict[str, Any]
    experience: list[ExperienceResponse]
    techStack: list[TechStackResponse]
    projects: list[ProjectResponse]
    publications: list[PublicationResponse]
    achievements: list[AchievementResponse]
    courses: list[CourseResponse]
    lifestyle: dict[str, Any]
    social: dict[str, Any]

File: apps/web/src/config/portfolio.ts
export const portfolioConfig = {
  hero: {
    name: "Your Name",
    title: "AI Engineer @ Your Company",
    tagline:
      "Building Agentic AI systems that automate complex workflows. Bridging cutting-edge GenAI with robust production systems.",
    avatar: "/images/avatar.jpg",
    location: "Ho Chi Minh City, Vietnam",
  },

  about: {
    summary:
      "Machine Learning Engineer with strong expertise in building Agentic AI systems and automating complex workflows. Proficient in orchestrating LLMs, implementing Human-in-the-loop architectures, and enforcing software engineering standards.",
    highlights: [
      "Architected Agentic AI systems using Pydantic-AI and OpenRouter",
      "Implemented Human-in-the-loop (HITL) capabilities for secure tool usage",
      "Published research on Object Detection at IEEE RIVF 2022",
      "IELTS 7.0 | GPA 8.42/10 - Top 10 graduating students at UIT",
    ],
  },

  experience: [
    {
      company: "Your Company",
      role: "AI Engineer",
      period: "Jun 2025 - Present",
      highlights: [
        "Architected Agentic AI system using Pydantic-AI to automate Customer Experience workflows",
        "Engineered dynamic tool execution with Human-in-the-loop capabilities",
        "Integrated Ragflow into Chatwoot for context-aware RAG support",
        "Orchestrated end-to-end automation for Ticket Exchange flows",
      ],
    },
    {
      company: "VUCAR",
      role: "Machine Learning Engineer",
      period: "Jan 2025 - Apr 2025",
      highlights: [
        "Developed VuSEO - AI-Agent automation tool processing 40-50 articles/day",
        "Enhanced ML model for car pricing accuracy",
      ],
    },
    {
      company: "UIT-Together Research Group",
      role: "Research Assistant",
      period: "Sep 2020 - Aug 2024",
      highlights: [
        "Researched Object Detection, Stable Diffusion, and Face Recognition",
        "Developed real-time demos and optimized inference using ONNX Runtime",
        "Mentored new research assistants on coding and communication",
      ],
    },
  ],

  education: {
    school: "VNU-HCM University of Information Technology",
    degree: "BS in Computer Science",
    period: "2020 - 2024",
    gpa: "8.42/10",
    rank: "Top 10/32 graduating students",
    coursework: [
      "Deep Learning Techniques",
      "Advanced Computer Vision",
      "Probability and Statistics",
      "Python for ML",
      "Data Mining",
    ],
  },

  techStack: [
    { name: "Python", icon: "python", category: "language" as const },
    { name: "TypeScript", icon: "typescript", category: "language" as const },
    { name: "Pydantic-AI", icon: "python", category: "ai" as const },
    { name: "PyTorch", icon: "pytorch", category: "ai" as const },
    { name: "FastAPI", icon: "fastapi", category: "backend" as const },
    { name: "Next.js", icon: "nextjs", category: "frontend" as const },
    { name: "PostgreSQL", icon: "postgres", category: "database" as const },
    { name: "Docker", icon: "docker", category: "devops" as const },
    { name: "Kubernetes", icon: "k8s", category: "devops" as const },
    { name: "Ragflow", icon: "ai", category: "ai" as const },
  ],

  projects: [
    {
      id: "1",
      title: "Agentic AI System @ Your Company",
      description:
        "Production system automating Customer Experience workflows with Human-in-the-loop tool execution and RAG integration",
      tags: ["Pydantic-AI", "OpenRouter", "Ragflow", "HITL"],
    },
    {
      id: "2",
      title: "VuSEO Automation",
      description:
        "AI-Agent powered SEO automation processing 40-50 articles/day with Google Workspace integration",
      tags: ["Python", "Strapi", "Google API", "AI Agent"],
    },
    {
      id: "3",
      title: "Homelab Infrastructure",
      description:
        "Self-hosted services on Kubernetes with monitoring via k9s, Kibana, and Logfire",
      tags: ["Docker", "K8s", "Terraform", "Monitoring"],
    },
  ],

  publications: [
    {
      title:
        "Empirical Study of the Performance of Object Detection Methods on Road Marking Dataset",
      venue: "IEEE RIVF 2022",
      doi: "10.1109/RIVF55975.2022.10013909",
      year: 2022,
    },
  ],

  achievements: [
    {
      title: "Innovation Award",
      event: "AI Tempo Run Competition",
      organization: "UIT AI Club",
      year: 2021,
    },
  ],

  courses: [
    {
      title: "Machine Learning Engineer K3",
      year: 2024,
      focus: ["Containerization", "Model Deployment", "CI/CD", "Data Engineering"],
    },
  ],

  lifestyle: {
    music: {
      instruments: ["Guitar", "Piano"],
      currentlyPlaying: "Lo-fi Beats",
    },
    routines: [
      "Morning: Code & Coffee",
      "Afternoon: Deep Work",
      "Evening: Music & Reading",
    ],
  },

  social: {
    github: "https://github.com/yourusername",
    linkedin: "https://www.linkedin.com/in/yourprofile/",
    email: "contact@yourdomain.com",
    phone: "+00 000 000 000",
  },
};

export type PortfolioConfig = typeof portfolioConfig;
